### DISCUSSION TITLE
Action used in another action

### DISCUSSION THREAD

**Discussion by varna** [mobxjs/mobx-state-tree#1873] [Q&A] [answered]
How can I use action in another action?
If I use `self`, then I get an error: `Property 'setViewState' does not exist on type`.
My "best" solution for a singleton model is:
```ts
import { types, getSnapshot, applySnapshot, SnapshotIn } from "mobx-state-tree"

// @ts-ignore
export const MapModel = types
.model("Map",{
  longitude: types.number,
  latitude: types.number,
  zoom: types.number
})
.views(self => ({
  get viewState() {
    return getSnapshot(self)
  }
}))
.actions(self => ({
  setViewState(snapshot: SnapshotIn<typeof MapModel>) {
    applySnapshot(self, snapshot)
  },
  fitBounds(/** ... */) {
    model.setViewState(/** ... */)
  }
}))

const model = MapModel.create({
  longitude: -122.41669,
  latitude: 37.7853,
  zoom: 13,
})

export default model
``` 

1. What should I do with a simple model?
2. I have the same problems with getters. i.e. `bbox` view used in a `fitBounds` action. Or creating getter for area as number and then another getter that uses that area to create a string with units.

**Comment by EmilTholin** [comment, marked as answer]
Hi @varna!

This is discussed in length in the [`TypeScript and MST` part of the documentation](https://mobx-state-tree.js.org/tips/typescript#typing-self-in-actions-and-views), and has a few solutions. I personally prefer splitting the `.action` block up into two separate blocks:

```ts
export const MapModel = types
  .model("Map", {
    longitude: types.number,
    latitude: types.number,
    zoom: types.number,
  })
  .views((self) => ({
    get viewState() {
      return getSnapshot(self);
    },
  }))
  .actions((self) => ({
    setViewState(snapshot: SnapshotIn<typeof MapModel>) {
      applySnapshot(self, snapshot);
    },
  }))
  .actions((self) => ({
    fitBounds(/** ... */) {
      self.setViewState(/** ... */);
    },
  }));
```

**Reply by varna** [reply]
Amazing. Sorry that I couldn't find that myself.

**Reply by EmilTholin** [reply]
No worries.
